### PR TITLE
👈 Add `tokenizer` arg back and add deprecation guidelines

### PR DESCRIPTION
**Pull request by qgallouedec**
# What does this PR do?

#2162 introduced a breaking change on all trainers except DPO and SFT (`tokenizer` arg replaced by `processing_class`. We've had feedback that this change was too abrupt, so we're reintroducing this argument with an extended timeline for its removal and also clarifying our removal strategy.

This PR will be part of a patch release for v0.12

Here is the proposed schedule depending of the usage for each trainer:

| Trainer       | Num models on the Hub | Argument removed in version |
| ------------- | --------------------- | ------------------- |
| SFT           | 15,097                 | 0.16                |
| DPO           | 3,053                 | 0.16                |
| PPO           | 441                   | 0.15                |
| ORPO          | 320                   | 0.15                |
| Reward        | 270                   | 0.15                |
| KTO           | 80                    | 0.14                |
| CPO           | 26                    | 0.14                |
| Online DPO    | 14                    | 0.14                |
| RLOO          | 4                     | 0.14                |
| XPO           | 3                     | 0.13                |
| Nash          | 1                     | 0.13                |
| GKD           | 1                     | 0.13                |
| Iterative SFT | 1                     | 0.13                |
| BCO           | 0                     | 0.13                |

cc @muellerzr 

Related: #2290 #2226 #2207 #2218

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a GitHub issue? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/trl/tree/main/docs).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Review by kashif — APPROVED** [review]

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -256,3 +256,30 @@ That's how `make test` is implemented (without the `pip install` line)!
 
 You can specify a smaller set of tests to test only the feature
 you're working on.
+
+### Deprecation and Backward Compatibility
```
Comment (lewtun) [inline CONTRIBUTING.md:260]: Super nice documentation!

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -256,3 +256,30 @@ That's how `make test` is implemented (without the `pip install` line)!
...
+   warnings.warn(
+       "The `Trainer.foo` method is deprecated and will be removed in version 0.14.0. "
+       "Please use the `Trainer.bar` class instead.",
+       FutureWarning,
+   )
+   ```
+
+The deprecation and removal schedule is based on each feature's usage and impact, with examples at two extremes:
+
+- **Experimental or Low-Use Features**: For a feature that is experimental or has limited usage, backward compatibility may not be maintained between releases. Users should therefore anticipate potential breaking changes from one version to the next.
```
Comment (lewtun) [inline CONTRIBUTING.md:281]: At some point, we could consider marking some features in the docs that we consider to be experimental, similar to how this is done in `transformers`: https://huggingface.co/docs/transformers/v4.46.2/en/main_classes/agent#agents--tools

**Review by lewtun — APPROVED** [review]
Thanks a lot for enabling backwards compatibility @qgallouedec! The PR LGTM and the new docs are really well written 🔥

File: CONTRIBUTING.md (hunk 2 above)
Comment (qgallouedec) [reply CONTRIBUTING.md:281]: Good point ! I'll add a first example for judges

Comment (qgallouedec) [reply CONTRIBUTING.md:281]: 3ce399c

Screenshot 2024-11-11 at 19 05 22

**Pull request merged**